### PR TITLE
ISSUE-2348 Implement UnauthenticatedBlobAccess/set create

### LINKED ISSUE #2348: Implement UnauthenticatedBlobAccess/set create
## Why
EPIC: https://github.com/linagora/tmail-backend/issues/2297
A JMAP method to grant blob access.

## How 

Sample request:

```json
[
  "UnauthenticatedBlobAccess/set",
  {
    "accountId": "ACCOUNT_ID",
    "create": {
      "BLOB_ID_1": {}
    }
  },
  "c1"
]
```

Sample response:

```json
[
  "UnauthenticatedBlobAccess/set",
  {
    "created": {
      "BLOB_ID_1": {
        "token": "TOKEN_VALUE"
      }
    }
  },
  "c1"
]
```

- Add the new capability object:
  ```json
  {
    "com:linagora:params:jmap:unauthenticated:blob:access": {
      "endpoint": "https://jmap.example.com/unauthenticatedDownload/{accountId}/{blobId}?token={token}",
      "tokenTtlInSeconds": 300
    }
  }
  ```

- Implement `UnauthenticatedBlobAccess/set create`
  Pay attention to not allow signing blob the user has no access to.

## DoD
Guice binding + JMAP contract tests for the method.

### PR DESCRIPTION
**Pull request by quantranhong1999**
resolve #2348

## Summary by CodeRabbit

* **New Features**
  * Added UnauthenticatedBlobAccess JMAP capability: sessions advertise a download endpoint and token TTL; clients can request short-lived download tokens with per-blob created/notCreated results.

* **Improvements**
  * Stricter blob access validation with resolver fallback to better authorize attachments and message parts.
  * Configurable unauthenticated-token storage choice (Redis, memory, or disabled) via configuration.

* **Tests**
  * New comprehensive integration tests (memory, distributed, Postgres) and shared contract covering creation, delegation, errors, mixed outcomes, limits, and token invalidation.

[Review Change Stack](https://app.coderabbit.ai/change-stack/linagora/tmail-backend/pull/2414?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: tmail-backend/integration-tests/jmap/distributed-jmap-integration-tests/src/test/java/com/linagora/tmail/james/DistributedUnauthenticatedBlobAccessSetMethodTest.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+        try {
+            Files.writeString(serverUtils.getConfigFolder().resolve("redis.properties"),
+                "redisURL=" + REDIS_EXTENSION.dockerRedis().redisURI() + "\n");
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+
+        return serverUtils.getConfigurationPath();
+    }
```
Comment (chibenwa) [inline tmail-backend/integration-tests/jmap/distributed-jmap-integration-tests/src/test/java/com/linagora/tmail/james/DistributedUnauthenticatedBlobAccessSetMethodTest.java]: Why is this needed? Why can't we instead inject a configuration POJO?

File: tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/UnauthenticatedBlobAccessSetMethodContract.scala [hunk 2]
```diff
@@ -0,0 +1,706 @@
...
+      .isEqualTo(
+        s"""{
+           |    "$blobId": {
+           |        "token": "$${json-unit.ignore}"
+           |    }
+           |}""".stripMargin)
+    assertThat(firstToken).isNotEqualTo(secondToken)
+    assertThat(tokenProbe(server).isValid(bobAccountId, blobId, firstToken)).isFalse
+    assertThat(tokenProbe(server).isValid(bobAccountId, blobId, secondToken)).isTrue
+  }
```
Comment (chibenwa) [inline tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/UnauthenticatedBlobAccessSetMethodContract.scala:506]: While I anm not opposed to it it seems a edge case given that we have a short expiry.

We can indeed keep it as is but it comes as a surprise to me.

File: tmail-backend/integration-tests/jmap/postgres-jmap-integration-tests/src/test/java/com/linagora/tmail/james/PostgresUnauthenticatedBlobAccessSetMethodTest.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+        try {
+            Files.writeString(serverUtils.getConfigFolder().resolve("redis.properties"),
+                "redisURL=" + REDIS_EXTENSION.dockerRedis().redisURI() + "\n");
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+
+        return serverUtils.getConfigurationPath();
+    }
```
Comment (chibenwa) [inline tmail-backend/integration-tests/jmap/postgres-jmap-integration-tests/src/test/java/com/linagora/tmail/james/PostgresUnauthenticatedBlobAccessSetMethodTest.java]: I'd prefer we inject a POJO

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/UnauthenticatedBlobAccessSetMethod.scala [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+      case jsObject: JsObject if jsObject.value.isEmpty => Right(())
+      case _: JsObject => Left(SetError.invalidArguments(SetErrorDescription("create value must be an empty object")))
+      case _ => Left(SetError.invalidArguments(SetErrorDescription("create value must be an empty object")))
+    }
+
+  private def parseBlobIds(blobIdAsString: String): Either[SetError, BlobIds] =
+    (for {
+      jmapBlobId <- JmapBlobId.of(blobIdAsString)
+      blobStoreBlobId <- Try(blobIdFactory.parse(blobIdAsString))
+    } yield BlobIds(jmapBlobId, blobStoreBlobId)) match {
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/UnauthenticatedBlobAccessSetMethod.scala]: This () is clever to avoid allocating a variable and write a one liner, that's true

However it felt hard to read and understand what goes on. I think allocating a `val` would yield something easier to read.

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/UnauthenticatedBlobAccessSetMethod.scala [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+    JavaAccountId.fromString(accountId.id.value)
+
+  private def closeBlobContentAsync(blob: Blob): Unit =
+    SMono.fromCallable(() => blob.content.close())
+      .subscribeOn(ReactorUtils.BLOCKING_CALL_WRAPPER)
+      .onErrorResume(e => {
+        LOGGER.warn("Failed to close resolved blob content", e)
+        SMono.empty[Unit]
+      })
+      .subscribe()
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/UnauthenticatedBlobAccessSetMethod.scala]: Why do we need this "close" to occur outside of the reactive chain?

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/UnauthenticatedBlobAccessSetMethod.scala [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+      .map(UnauthenticatedBlobAccessCreationResults)
+
+  private def create(mailboxSession: MailboxSession,
+                     accountId: AccountId,
+                     blobIdAsString: String,
+                     createValue: JsValue): SMono[UnauthenticatedBlobAccessCreationResult] =
+    validateCreateEntry(blobIdAsString, createValue) match {
+      case Left(error) => SMono.just(UnauthenticatedBlobAccessCreationFailure(blobIdAsString, error))
+      case Right(blobIds) => blobResolvers.resolve(blobIds.jmapBlobId, mailboxSession)
+        .flatMap(blob => SMono(tokenRepository.generate(toJavaAccountId(accountId), blobIds.blobStoreBlobId))
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/UnauthenticatedBlobAccessSetMethod.scala]: Forgive me if I am wrong but this ends up reading the actual contents.

Pretty sure that this will end up doing requests to S3, and for some implementation (mailbox messages) read fully the S3 data.

I understand we want to hook on BlobResolver modularity that is convenient but shouldn't we have a `BlobResolver::validateAccess` that actually returns a boolean directly ?

**Comment by Arsnael** [comment]
Docker image published for this PR: linagora/tmail-backend-pr:2414

**Review by Arsnael — COMMENTED** [review]
Read it nothing to add

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/UnauthenticatedBlobAccessSetMethod.scala (hunk 5 above)
Comment (quantranhong1999) [reply tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/UnauthenticatedBlobAccessSetMethod.scala]: The idea was to close the blob async to not delay the generated token returned. But if we attempt to have `BlobResolver::validateAccess` below, we should not need to `close` anymore.

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/UnauthenticatedBlobAccessSetMethod.scala (hunk 6 above)
Comment (quantranhong1999) [reply tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/UnauthenticatedBlobAccessSetMethod.scala]: > I understand we want to hook on BlobResolver modularity that is convenient but shouldn't we have a BlobResolver::validateAccess that actually returns a boolean directly

Let me try.

Comment (quantranhong1999) [reply tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/UnauthenticatedBlobAccessSetMethod.scala]: James changes: https://github.com/apache/james-project/pull/3051

Comment (chibenwa) [reply tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/UnauthenticatedBlobAccessSetMethod.scala]: Very cool, thank you!

File: tmail-backend/integration-tests/jmap/distributed-jmap-integration-tests/src/test/java/com/linagora/tmail/james/DistributedUnauthenticatedBlobAccessSetMethodTest.java (hunk 1 above)
Comment (chibenwa) [reply tmail-backend/integration-tests/jmap/distributed-jmap-integration-tests/src/test/java/com/linagora/tmail/james/DistributedUnauthenticatedBlobAccessSetMethodTest.java]: > Why is this needed? Why can't we instead inject a configuration POJO?

Don't forget to address eventually this comment please

Comment (quantranhong1999) [reply tmail-backend/integration-tests/jmap/distributed-jmap-integration-tests/src/test/java/com/linagora/tmail/james/DistributedUnauthenticatedBlobAccessSetMethodTest.java]: Yes, of course. I will. In the later batch commits.

**Review by chibenwa — APPROVED** [review]

**Comment by Arsnael** [comment]
Docker image published for this PR: linagora/tmail-backend-pr:2414

**Review by Arsnael — APPROVED** [review]

**Comment by Arsnael** [comment]
Complementary PR on james merged, please rebase and update james submodule

**Comment by quantranhong1999** [comment]
> Complementary PR on james merged, please rebase and update james submodule

Rebased.

**Comment by Arsnael** [comment]
Docker image published for this PR: linagora/tmail-backend-pr:2414
